velox/functions/sparksql/String.h

```diff
@@ -583,4 +583,83 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
   }
 };
 
+/// translate(string, matchStr, replaceStr) -> varchar
+///
+///   Returns a new translated string. It translates the character in ``string``
+///   by a character in ``replaceStr``. The character in ``replaceStr`` is
+///   corresponding to the character in ``matchStr``. The translation will
+///   happen when any character in ``string`` matching with the character in the
+///   ``matchStr``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  using MapType = std::unordered_map<StringView, StringView>;
+  MapType dict_;
+  bool isConstantTranslateDict = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDict(
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    MapType dict;
+    int i = 0, j = 0;
+    while (i < matchStr.size()) {
+      StringView replacePiece;
+      if (j >= replaceStr.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = utf8proc_char_length(replaceStr.data() + j);
+        replacePiece = StringView(replaceStr.data() + j, replaceCharCount);
+        j = j + replaceCharCount;
+      }
+      int matchCharCount = utf8proc_char_length(matchStr.data() + i);
+      StringView matchPiece = StringView(matchStr.data() + i, matchCharCount);
+      dict.try_emplace(matchPiece, replacePiece);
+      i = i + matchCharCount;
+    }
+    return dict;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* matchStr,
+      const arg_type<Varchar>* replaceStr) {
+    if (matchStr != nullptr && replaceStr != nullptr) {
+      dict_ = buildDict(*matchStr, *replaceStr);
+      isConstantTranslateDict = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& matchStr,
+      const arg_type<Varchar>& replaceStr) {
+    if (!isConstantTranslateDict) {
+      dict_ = buildDict(matchStr, replaceStr);
+    }
+    // No need to do the replacement.
+    if (dict_.empty()) {
+      result.setNoCopy(StringView(input.data(), input.size()));
```

Comment:
Why setNoCopy? I imagine we need to copy the string, no?

Reply:
For this case, just re-use the input string as result. So I use setNoCopy. Am I missing something? Thanks!

Reply:
To reuse input strings, one needs to set reuse_strings_from_arg to tell the framework that input string buffers need to be added to the result vector. Otherwise, you'll end up having a result vector referring to memory that it doesn't own.

https://facebookincubator.github.io/velox/develop/scalar-functions.html#zero-copy-string-result

Reply:
Thanks for letting me know this. Yes, I missed this key point. Just added the reuse flag.
A question. For such function that produces no-copy result for some cases and doesn't for other cases, is it a best practice to still set the reuse flag? As with it, the framework will not let the input buffer go away prematurely, even for copy cases.

Reply:
In this case, I think it would be rare that input strings can be reused. Hence, I'd suggest not to set reuse flag and not to use setNoCopy. Otherwise. you'll end up keeping unused strings buffers in memory for longer than needed and waste memory.

Reply:
Absolutely agree. That reuse case is very rare in real workload. I just removed the usage of `setNoCopy`, as well as the reuse flag. Thanks!